File: internal/diagnostics/diagnostics.go
package diagnostics

import (
	"fmt"
	"os/exec"
	"strings"
)

var allowedCommands = map[string]bool{
	"ping":       true,
	"nc":         true,
	"netstat":    true,
	"ss":         true,
	"curl":       true,
	"traceroute": true,
	"nslookup":   true,
	"dig":        true,
	"ps":         true,
	"lsof":       true,
	"df":         true,
	"free":       true,
	"dpkg":       true,
	"apt":        true,
	"rpm":        true,
	"cat":        true,
	"ufw":        true,
	"iptables":   true,
	"systemctl":  true,
	"wget":       true,
}

// IsAllowed checks if a command is safe to execute as a diagnostic.
func IsAllowed(cmd string) bool {
	parts := strings.Fields(cmd)
	if len(parts) == 0 {
		return false
	}

	base := parts[0]
	if base == "sudo" && len(parts) > 1 {
		return false
	}

	if !allowedCommands[base] {
		return false
	}

	switch base {
	case "systemctl":
		if len(parts) < 2 || parts[1] != "status" {
			return false
		}
	case "ufw":
		if len(parts) < 2 || parts[1] != "status" {
			return false
		}
	case "iptables":
		if !containsFlag(parts, "-L") {
			return false
		}
	case "cat":
		if len(parts) < 2 || !strings.HasPrefix(parts[len(parts)-1], "/etc/") {
			return false
		}
	case "wget":
		if containsFlag(parts, "-O") {
			return false
		}
	}

	return true
}

// RunDiagnostic executes a single diagnostic command and returns its output.
func RunDiagnostic(cmd string) (string, error) {
	if !IsAllowed(cmd) {
		return "", fmt.Errorf("command not allowed: %q", cmd)
	}

	out, err := exec.Command("bash", "-c", cmd).CombinedOutput()
	output := string(out)
	if err != nil {
		return output, nil
	}
	return output, nil
}

// RunAll executes multiple diagnostic commands and returns combined output.
func RunAll(commands []string) map[string]string {
	results := make(map[string]string)
	for _, cmd := range commands {
		output, err := RunDiagnostic(cmd)
		if err != nil {
			results[cmd] = fmt.Sprintf("[BLOCKED] %s", err.Error())
		} else {
			results[cmd] = output
		}
	}
	return results
}

func containsFlag(parts []string, flag string) bool {
	for _, p := range parts {
		if p == flag {
			return true
		}
	}
	return false
}

File: nr_diagnose/diagnostics.py
"""Safe diagnostic command execution with allowlist."""

import subprocess
from typing import Dict, List, Optional, Tuple

ALLOWED_COMMANDS = {
    "ping", "nc", "netstat", "ss", "curl", "traceroute",
    "nslookup", "dig", "ps", "lsof", "df", "free",
    "dpkg", "apt", "rpm", "cat", "ufw", "iptables",
    "systemctl", "wget", "tnsping", "echo",
}


def is_allowed(cmd: str) -> bool:
    """Check if a command is safe to execute as a diagnostic."""
    parts = cmd.split()
    if not parts:
        return False

    base = parts[0]

    # No sudo
    if base == "sudo":
        return False

    if base not in ALLOWED_COMMANDS:
        return False

    # Additional restrictions per command
    if base == "systemctl":
        if len(parts) < 2 or parts[1] != "status":
            return False
    elif base == "ufw":
        if len(parts) < 2 or parts[1] != "status":
            return False
    elif base == "iptables":
        if "-L" not in parts:
            return False
    elif base == "cat":
        if len(parts) < 2 or not parts[-1].startswith("/etc/"):
            return False
    elif base == "wget":
        if "-O" in parts:
            return False

    return True


def run_diagnostic(cmd: str) -> Tuple[str, Optional[str]]:
    """Execute a single diagnostic command and return (output, error_msg).

    Returns (output, None) on success or blocked output,
    (error_message, error_message) if not allowed.
    """
    if not is_allowed(cmd):
        return f"[BLOCKED] command not allowed: {cmd!r}", f"command not allowed: {cmd!r}"

    try:
        result = subprocess.run(
            ["bash", "-c", cmd],
            capture_output=True, text=True, timeout=30
        )
        # Return combined output regardless of exit code
        output = result.stdout + result.stderr
        return output.strip(), None
    except subprocess.TimeoutExpired:
        return "[TIMEOUT] command timed out after 30s", None
    except Exception as e:
        return f"[ERROR] {e}", None


def run_all(commands: List[str]) -> Dict[str, str]:
    """Execute multiple diagnostic commands and return combined output."""
    results: Dict[str, str] = {}
    for cmd in commands:
        output, error = run_diagnostic(cmd)
        if error:
            results[cmd] = f"[BLOCKED] {error}"
        else:
            results[cmd] = output
    return results
